In [15]:
import pandas as pd
import numpy as np
import os

### Preparing Paths to feed Images to Neural Net:

Now we finally have all of the hard work done. During this process Iwas able to create 43439 unique images to feed to my neural network. This is an incredibly small dataset to train a neural net but, we'll see what kind of results we can get!

In order to create an easy to read table of points to feed to the model we're first going to read in the Folder titles (Which are also the same as the title of the Wikipedia article being read).

In [16]:
english_folders = os.listdir('./data/audio_images/English/english')
german_folders = os.listdir('./data/audio_images/German/german')

### English Files:

I'm going to create the DataFrame with 3 labels. The image, The title it belongs to, and the language it is in.

The loop below pulls all the image file names and fills out a list in order. It then generates a list of the same length with the title names. They both get added to an overall list and the over all length is saved it a length variable ```lens```

In [17]:
file_list = []
names_list = []
lens = 0
for i in english_folders:
    files = os.listdir(f'./data/audio_images/English/english/{i}')
    names = [i] * len(files)
    file_list += files
    names_list += names
    lens += (len(files))
np.asarray(file_list).shape

(23232,)

After we check the shapes of the two lists to make sure they are the same length. If there were any errors in the above script the lengths would be different.

In [18]:
np.asarray(names_list).shape

(23232,)

The lengths are the same which means our lists were made correctly. 

Now I generate a list of ones (my binary indicator for English) and we put them all in a DataFrame

In [19]:
language = np.ones(lens)

In [20]:
df_english = pd.DataFrame()
df_english['Files'] = file_list
df_english['Names'] = names_list
df_english['Language'] = language

df_english

Files                    Names  Language
0         0_0-15_audio_to_img.jpg  Longest_word_in_English       1.0
1      1_105-120_audio_to_img.jpg  Equal_Protection_Clause       1.0
2         1_0-15_audio_to_img.jpg  Equal_Protection_Clause       1.0
3      1_120-135_audio_to_img.jpg  Equal_Protection_Clause       1.0
4        1_60-75_audio_to_img.jpg  Equal_Protection_Clause       1.0
...                           ...                      ...       ...
23227  0_120-135_audio_to_img.jpg          Richard_Branson       1.0
23228    0_15-30_audio_to_img.jpg          Richard_Branson       1.0
23229  0_285-300_audio_to_img.jpg          Richard_Branson       1.0
23230    0_75-90_audio_to_img.jpg          Richard_Branson       1.0
23231  0_105-120_audio_to_img.jpg          Richard_Branson       1.0

[23232 rows x 3 columns]

### German:

Rinse and Repeat:

Only difference here is we use the German file path ```files``` and we use zeros instead of ones.

In [21]:
file_list = []
names_list = []
lens = 0
for i in german_folders:
    files = os.listdir(f'./data/audio_images/German/german/{i}')
    names = [i] * len(files)
    file_list += files
    names_list += names
    lens += (len(files))
np.asarray(file_list).shape

(20207,)

In [22]:
language = np.zeros(lens)

In [23]:
df_german = pd.DataFrame()
df_german['Files'] = file_list
df_german['Names'] = names_list
df_german['Language'] = language

df_german

Files                Names  Language
0      0_225-240_audio_to_img.jpg            Rhodopsin       0.0
1       0_90-105_audio_to_img.jpg            Rhodopsin       0.0
2      0_135-150_audio_to_img.jpg            Rhodopsin       0.0
3      0_180-195_audio_to_img.jpg            Rhodopsin       0.0
4      0_240-255_audio_to_img.jpg            Rhodopsin       0.0
...                           ...                  ...       ...
20202  0_120-135_audio_to_img.jpg  Deutscher_Bundestag       0.0
20203    0_15-30_audio_to_img.jpg  Deutscher_Bundestag       0.0
20204  0_285-300_audio_to_img.jpg  Deutscher_Bundestag       0.0
20205    0_75-90_audio_to_img.jpg  Deutscher_Bundestag       0.0
20206  0_105-120_audio_to_img.jpg  Deutscher_Bundestag       0.0

[20207 rows x 3 columns]

### Totals:

In total we have 23232 English Files and 20207 German Files to work with

### Bring them together:

We're going to concatenate the DataFrames and reset the index

In [24]:
df = pd.concat([df_english, df_german])

In [25]:
df.reset_index(inplace=True)
df

index                       Files                    Names  Language
0          0     0_0-15_audio_to_img.jpg  Longest_word_in_English       1.0
1          1  1_105-120_audio_to_img.jpg  Equal_Protection_Clause       1.0
2          2     1_0-15_audio_to_img.jpg  Equal_Protection_Clause       1.0
3          3  1_120-135_audio_to_img.jpg  Equal_Protection_Clause       1.0
4          4    1_60-75_audio_to_img.jpg  Equal_Protection_Clause       1.0
...      ...                         ...                      ...       ...
43434  20202  0_120-135_audio_to_img.jpg      Deutscher_Bundestag       0.0
43435  20203    0_15-30_audio_to_img.jpg      Deutscher_Bundestag       0.0
43436  20204  0_285-300_audio_to_img.jpg      Deutscher_Bundestag       0.0
43437  20205    0_75-90_audio_to_img.jpg      Deutscher_Bundestag       0.0
43438  20206  0_105-120_audio_to_img.jpg      Deutscher_Bundestag       0.0

[43439 rows x 4 columns]

### Save to CSV:

The last thing we need to do is drop the index column and make a new column with the full path for the model to read.

Our total dataset is 43439 rows so we're going to save this to CSV so we can reference these values when building the model.

In [31]:
df.drop('index', axis=1, inplace=True)
df

KeyError: "['index'] not found in axis"

In [32]:
df['FilePath'] = ''
for i, j in df.iterrows():
    if df['Language'][i] == 1:
        path = './data/audio_images/English/english'
        df['FilePath'][i] = f'{path}/{df["Names"][i]}/{df["Files"][i]}'
    else:
        path = './data/audio_images/German/german'
        df['FilePath'][i] = f'{path}/{df["Names"][i]}/{df["Files"][i]}'
    

/home/houston/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
/home/houston/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [33]:
df

Files                    Names  Language  \
0         0_0-15_audio_to_img.jpg  Longest_word_in_English       1.0   
1      1_105-120_audio_to_img.jpg  Equal_Protection_Clause       1.0   
2         1_0-15_audio_to_img.jpg  Equal_Protection_Clause       1.0   
3      1_120-135_audio_to_img.jpg  Equal_Protection_Clause       1.0   
4        1_60-75_audio_to_img.jpg  Equal_Protection_Clause       1.0   
...                           ...                      ...       ...   
43434  0_120-135_audio_to_img.jpg      Deutscher_Bundestag       0.0   
43435    0_15-30_audio_to_img.jpg      Deutscher_Bundestag       0.0   
43436  0_285-300_audio_to_img.jpg      Deutscher_Bundestag       0.0   
43437    0_75-90_audio_to_img.jpg      Deutscher_Bundestag       0.0   
43438  0_105-120_audio_to_img.jpg      Deutscher_Bundestag       0.0   

                                                FilePath  
0      ./data/audio_images/English/english/Longest_wo...  
1      ./data/audio_images/English/english/Equal_Prot...  
2      ./data/audio_images/English/english/Equal_Prot...  
3      ./data/audio_images/English/english/Equal_Prot...  
4      ./data/audio_images/English/english/Equal_Prot...  
...                                                  ...  
43434  ./data/audio_images/German/german/Deutscher_Bu...  
43435  ./data/audio_images/German/german/Deutscher_Bu...  
43436  ./data/audio_images/German/german/Deutscher_Bu...  
43437  ./data/audio_images/German/german/Deutscher_Bu...  
43438  ./data/audio_images/German/german/Deutscher_Bu...  

[43439 rows x 4 columns]

In [34]:
df.to_csv('./data/feed_dict.csv', index=False)